lru_ttl.go

```diff
@@ -0,0 +1,170 @@
+package lru
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/golang-lru/simplelru"
+)
+
+// NOTE: this implementation will ensure that the cache will become eventually consistent.
+// Expired items will stay in the cache until it is removed.
+//
+// When a GET is received on the expired item, the item is removed as part of the GET
+// call. But, the other functions would still include the expired item in their result until
+// it is removed by the cleanup routine.
+//
+// `Add` is the only call which will update the lastAccessTime of an item.
+
+// CacheWithTTL implements thread safe fixed size LRU cache with TTL
+type CacheWithTTL struct {
+	*simplelru.LRU
+	lock sync.RWMutex
+	TTL  time.Duration
+}
+
+// cacheValue is a wrapper around the cache value to hold last accessed time
+type cacheValue struct {
+	value          interface{}
+	lastAccessTime time.Time
+}
+
+// NewTTL constructs an LRU of the given size with the given TTL
+func NewTTL(size int, ttl time.Duration) (simplelru.LRUCache, error) {
+	return NewTTLWithEvict(size, ttl, nil)
+}
+
+// NewTTLWithEvict constructs an LRU of the given size with given TTL
+// Also, sets up the evict function
+func NewTTLWithEvict(size int, ttl time.Duration, onEvict simplelru.EvictCallback) (simplelru.LRUCache, error) {
+	if size <= 0 {
+		return nil, errors.New("Must provide a positive size")
+	}
+
+	lru, err := simplelru.NewLRU(size,
+		func(k interface{}, v interface{}) {
+			if onEvict != nil {
+				onEvict(k, v.(cacheValue).value)
+			}
+		})
+	if err != nil {
+		return nil, err
+	}
+
+	lruWithTTL := &CacheWithTTL{LRU: lru, TTL: ttl}
+
+	// clean expired items
+	go lruWithTTL.cleanup()
```

Comment:
I would be hesitant to use anything that cannot be shutdown.  As implemented, it's a memory leak.